source/common/common/envoy_defaults.h

```diff
@@ -0,0 +1,131 @@
+
+#define PROTOBUF_GET_NUMBER_OR_PROFILE_DEFAULT(message, config, field_name, default_value)         \
+  ([&](const auto& msg) -> double {                                                                \
+    if (msg.has_##field_name()) {                                                                  \
+      return msg.field_name().value();                                                             \
+    }                                                                                              \
+    return DefaultsProfile::get().getNumber(config, #field_name, default_value);                   \
+  }((message)))
+
+#define PROTOBUF_GET_MS_OR_PROFILE_DEFAULT(message, config, field_name, default_value)             \
+  ([&](const auto& msg) -> std::chrono::milliseconds {                                             \
+    if (msg.has_##field_name()) {                                                                  \
+      return std::chrono::milliseconds(DurationUtil::durationToMilliseconds(msg.field_name()));    \
+    }                                                                                              \
+    return DefaultsProfile::get().getMs(config, #field_name, default_value);                       \
+  }((message)))
+
+#define PROTOBUF_GET_BOOL_OR_PROFILE_DEFAULT(message, config, field_name, default_value)           \
+  ([&](const auto& msg) -> bool {                                                                  \
+    if (msg.has_##field_name()) {                                                                  \
+      return msg.field_name().value();                                                             \
+    }                                                                                              \
+    return DefaultsProfile::get().getBool(config, #field_name, default_value);                     \
+  }((message)))
+
+namespace Envoy {
+
+class DefaultsProfile {
+public:
+  enum Profile {
+    Performant,
+    Safe,
+  };
+
+  /**
+   * ConfigContext encapsulates the full name of the config Protobuf message containing the field(s)
+   * for which defaults are being accessed.
+   */
+  class ConfigContext {
+  public:
+    ConfigContext() = default;
+    ConfigContext(const ConfigContext& context) = default;
+    ConfigContext(const Protobuf::Message& config) : ctx_(config.GetDescriptor()->full_name()) {}
+    ConfigContext(absl::string_view context) : ctx_(context) {}
+    const absl::string_view getContext() const { return ctx_; }
+    ConfigContext& appendField(absl::string_view field) {
+      absl::StrAppend(&ctx_, ".", field);
+      return *this;
+    }
+
+  private:
+    std::string ctx_;
+  };
+
+  /**
+   * DefaultsProfile intended to be constructed as a DefaultsProfileSingleton, so that default
+   * values can be accessed globally.
+   * @param profile enum value specifying which profile to load
+   */
+  DefaultsProfile(Profile profile = Performant);
```

Comment:
When does this get instantiated in Safe mode?

Reply:
Check out `source/server/server.cc:355`. Here is where the DefaultsProfile constructor is invoked for Envoy; it takes its `Profile` argument from the command line options.